sharing_portal/trovi.py

```diff
@@ -0,0 +1,273 @@
+
+
+class TroviException(Exception):
+    pass
+
+
+def url_with_token(path, token):
+    req = requests.PreparedRequest()
+    req.prepare_url(
+        urljoin(os.getenv("TROVI_API_BASE_URL"), path), dict(access_token=token)
+    )
+    return req.url
+
+
+def check_status(response, code):
+    if response.status_code != code:
+        print(response.text)
+        try:
+            response_json = response.json()
+            detail = response_json.get(
+                "detail", response_json.get("error_description", response.text)
+            )
+        except requests.exceptions.JSONDecodeError:
+            if response.status_code == 500:
+                detail = ""
+            else:
+                detail = response.text
+        request = response.request
+        request_path = urlparse(request.url).path
+        message = (
+            f"{request.method} {request_path} {response.status_code} "
+            f"returned, expected {code}: {detail}"
+        )
+        raise TroviException(message)
+
+
+def get_client_admin_token():
+    keycloak_client = KeycloakClient()
+    realm = KeycloakRealm(server_url=keycloak_client.server_url, realm_name="master")
+    openid = realm.open_id_connect(
+        client_id=keycloak_client.client_id, client_secret=keycloak_client.client_secret
+    )
+    creds = openid.client_credentials()
+    return get_token(creds["access_token"], is_admin=True)["access_token"]
+
+
+def get_token(token, is_admin=False):
+    scope = "artifacts:read artifacts:write"
+    if is_admin:
+        scope += " trovi:admin"
+    res = requests.post(
+        urljoin(os.getenv("TROVI_API_BASE_URL"), "/token/"),
+        json={
+            "grant_type": "token_exchange",
+            "subject_token": token,
+            "subject_token_type": "urn:ietf:params:oauth:token-type:jwt",
+            "scope": scope,
+        },
+    )
+    check_status(res, requests.codes.created)
+    return res.json()
+
+
+def list_artifacts(token, sort_by="updated_at"):
+    req = requests.PreparedRequest()
+    req.prepare_url(url_with_token("/artifacts/", token), dict(sort_by=sort_by))
+    res = requests.get(req.url)
+    check_status(res, requests.codes.ok)
+    LOG.info(res.json())
+    return res.json()["artifacts"]
+
+
+def list_tags(token):
+    res = requests.get(url_with_token("/meta/tags/", token))
+    check_status(res, requests.codes.ok)
+    return res.json()["tags"]
+
+
+def create_tag(token, tag):
+    json_data = {"tag": tag}
+    res = requests.post(url_with_token("/meta/tags/", token), json=json_data)
+    check_status(res, requests.codes.created)
+    return res.json()["tag"]
+
+
+def get_trovi_uuid(artifact_id):
+    artifact = Artifact.objects.get(pk=artifact_id)
+    return artifact.trovi_uuid
+
+
+def get_author(author, prompt_input=True):
+    matching_users = list(
+        User.objects.annotate(full_name=Concat("first_name", "last_name"))
+        .filter(full_name=author.name.replace(" ", ""))
```

Comment:
heh, clever. You could also do this to avoid the space replacement:

```python
User.objects.annotate(full_name=Concat("first_name", Value(" "), "last_name"))
```

Reply:
I noticed that as well in the data. I think in most cases it was trailing or leading whitespace. Maybe `author.name.trim()`?

Reply:
Sure, I can replace this code.